### DISCUSSION TITLE
Multi-language support

### DISCUSSION THREAD

**Discussion by andriinuts** [medusajs/medusa#2003] [Feature Requests]
Comment from @olivermrbl:
### What
We've included Multi-language support in our roadmap and will likely start scoping mid-next year. 

------
**OP**
Hello, would be nice to have multi-language support. It will make it much easier to manage content for multi-language stores.

**Comment by ppsirius** [comment]
Is there any plan for that in the roadmap?

**Comment by srindom** [comment]
Users with multi-language needs generally manage this in a CMS (e.g., Contentful), allowing Medusa only to hold commerce-related data. I like this approach as it doesn't clutter Medusa with data only relevant to the presentation layer. Nevertheless, I would love to explore the options for multi-language support. 

Could you add details about why you need this feature in Medusa and not another place in your stack? Would also be great to hear details on how you want the feature to work. E.g., which fields should be translatable, how should different locales be managed, etc.?

**Comment by ppsirius** [comment]
You could easily pass the translation with i18n and keep it in the repository. Translation of buttons or different constant layout elements does not really need CMS. But that solution is mainly for admin panel, in storefront you should get each translation also for different products so it is more complex rather translating a layout but is possible. We should also think about generated paths for SEO. 
Translation could have a huge impact on the adoption of medusa on the market because Shopify or Prestashop give translation out of the box and developers or business pick this solution with less work.

**Reply by srindom** [reply]
Just to clarify: I am referring to managing translations of things like product names.

Admin translations should be relatively straightforward, although it will require some work to implement an i18n library :)

**Comment by speciale** [comment]
I could help translating in French and Spanish. Would LOVE to!

**Comment by olivermrbl** [comment]
As @srindom mentions, there are different multi-language requests in play here. 

Regarding translating the admin system, we could proceed without doing much technical design. The only decision we have to make is around what i18n library to use. I'll let @kasperkristensen and @StephixOne make that decision, as they are leading the project's development. 

When that has been decided, we would highly appreciate your help with the migration to using the library as well as supporting more languages.

I'll make sure to keep this thread updated regarding our decision.

**Reply by popowych** [reply]
Thanks for your work and the energy you put in it. 
I'm sorry if I didn't get it right, but is there a real use case of translating the admin system. I mean the back office, is there really someone that needs to have the back office translated? I do guess you're not using medusa if you don't read/speak english. 
On the contrary, translating the content your customers actually see is really important. Otherwise, they won't shop your brand. No sales, no needs of an eshop. We'll implement our own solution to give that option to our customers. I think that feature would be great in the initial medusa package. 

From France, selling in Europe and beyond.

**Reply by speciale** [reply]
> I mean the back office, is there really someone that needs to have the back office translated? I do guess you're not using medusa if you don't read/speak english. 
> 
That's exactly my actual situation. I did translate the front. But I'm not the seller, I'm only the developer. The seller wants (prefers) a french version. Is the old container content question.

**Reply by dludemann** [reply]
>  but is there a real use case of translating the admin system. 

This would be useful in my scenario. I create e-commerce sites for clients, and would love to move them to medusa. I have several clients who do not speak english as a first language, but they sell on the American / UK markets. So the frontend has to be in english, but the backend would be great if it were their native language. 

It would be faster for them to be able to add/edit products or update inventories, see orders, make gift cards, etc. I would like to be able to hand off the website to them so they can maintain their orders versus me having to do it and then translate to them what is going on. 

This is actually the only reason that keeps me from rolling out medusa sites to all my clients, and why I stick to other e-commerce solutions for the time being.

**Reply by mariusa** [reply]
Why isn't Chrome browser translate good enough?

**Reply by dludemann** [reply]
> Why isn't Chrome browser translate good enough?

If you speak two languages and have ever tried using google browser translate, you know that it is not accurate. I personally speak english and spanish fluently and have done comparisons and it just doesn't translate correctly all the time and doesn't always catch the meaning. Google translate is a very "literal" translation.

**Reply by mariusa** [reply]
For now, but it is improving. It certainly removes a potential blocker, even if not perfectly. There are other blockers on the roadmap for which there are no workarounds.

**Reply by dludemann** [reply]
> For now, but it is improving. It certainly removes a potential blocker, even if not perfectly. There are other blockers on the roadmap for which there are no workarounds.

For you may not be a blocker. For some clients - who pay a lot of money to have a developer develop an e-commerce solution for them - "even if not perfectly" will not cut it. This is most certainly a blocker for a good amount of people and developers and there is no workaround but to use a competitor. 

The developers should tackle their roadmap as they see fit, and I'm sure there are other priorities. But whenever this becomes an issue that they would like to tackle, this discussion will help developers with ideas, feedback, and knowledge of how translation would likely be used by users.

**Reply by SGFGOV** [reply]
A very simple way is to use strapis i18 feature which I have implemented in my strapi repo.. you can query strapi for whenever you need a specific text.. Ideally the translation layer is a part of the presentation layer not a part of the backend or business logic layer.. now you can use the strapi repo repo or your own version..of strapi or another cms

**Reply by MrCasquette** [reply]
> As @srindom mentions, there are different multi-language requests in play here.
> 
> Regarding translating the admin system, we could proceed without doing much technical design. The only decision we have to make is around what i18n library to use. I'll let @kasperkristensen and @StephixOne make that decision, as they are leading the project's development.
> 
> When that has been decided, we would highly appreciate your help with the migration to using the library as well as supporting more languages.
> 
> I'll make sure to keep this thread updated regarding our decision.

Just in case I already use this for Directus UI French Translation:

https://crowdin.com/

It's a very great tools and there's (I think) a free version for open source projects :)

Plus users do not need to edit JSON but there's a cool dashboard for selecting translations and translate it.
And it's synced with the project

**Reply by SalahAdDin** [reply]
> > As @srindom mentions, there are different multi-language requests in play here.
> > Regarding translating the admin system, we could proceed without doing much technical design. The only decision we have to make is around what i18n library to use. I'll let @kasperkristensen and @StephixOne make that decision, as they are leading the project's development.
> > When that has been decided, we would highly appreciate your help with the migration to using the library as well as supporting more languages.
> > I'll make sure to keep this thread updated regarding our decision.
> 
> Just in case I already use this for Directus UI French Translation:
> 
> https://crowdin.com/
> 
> It's a very great tools and there's (I think) a free version for open source projects :)
> 
> Plus users do not need to edit JSON but there's a cool dashboard for selecting translations and translate it. And it's synced with the project
Wait a minute, does not Directus have internationalization?

**Comment by speciale** [comment]
Ok. I'm new to the Node universe and I was looking for the possibilities. I saw that there's an i18n and an i18next. If I understood, the second differs as she gives the translation option. Thanks for answering.

**Comment by adrien2p** [comment]
Just for ref https://github.com/medusajs/admin/issues/405#issue-1149918432

**Comment by SaadBazaz** [comment]
I like the way Django Admin and Django-SHOP manage it. They have certain fields marked as "translatable fields", then at runtime they use the Content-Language header to determine which language to use to choose the translatable field. This could all be done in Medusa Admin, with some minor UI changes. Although, you might need an additional database table to store translations 
@olivermrbl

**Reply by herawais** [reply]
Odoo also work this way

**Comment by dfernandesbsolus** [comment]
Vendure have a very good way to make this:

https://github.com/vendure-ecommerce/vendure/blob/master/packages/core/src/entity/product/product.entity.ts
https://github.com/vendure-ecommerce/vendure/blob/master/packages/core/src/entity/product/product-translation.entity.ts

**Reply by SaadBazaz** [reply]
Yeap. That's a good way to do it.

**Reply by dfernandesbsolus** [reply]
I started doing this, but I don't have time to finish this now.

```js
import { Column, Entity, Index, ManyToOne } from 'typeorm';
import { TranslationEntity } from 'typeorm-translatable';
import { Product } from './product';

@Entity()
export class ProductTranslation extends TranslationEntity<Product> {
  @Column()
  title?: string;

  @Column('text')
  subtitle?: string;

  @Column('text')
  description?: string;

  @Index({ unique: true, where: "deleted_at IS NULL" })
  @Column({ type: "text", nullable: true })
  handle: string | null

  @ManyToOne(
    () => Product,
    product => product.translations
  )
  source?: Product | undefined;
}
```

**Reply by MDmubarak786** [reply]
is it available in V2?

**Reply by SalahAdDin** [reply]
Is it finished? is there any plugin?

**Comment by einar-hjortdal** [comment]
Multi-language support is often essential, particularly in Europe. I believe taking the approach Vendure has taken is a good option.

I disagree on relying on a third-party CMS as a solution. I can see using Strapi as a solution for blogging, or to manage pages, but not to mirror data structures in the Medusa database.

On the other hand, I do not care for admin frontend translations but if the capability is provided the community will take care of those. What I am concerned about is exclusively product, shipping and payment-related data that is to be displayed to customers.

22/04 update:
I have looked at the sqlite database schema created by medusa, including inventory and stock location modules (I do not know TypeORM so I can only analyze the raw schema).

The changes I would do:
1) Create a table containing locale codes
  ```sql
  CREATE TABLE "locale" (
      "id" VARCHAR PRIMARY KEY NOT NULL, -- generate entity id
      "code" VARCHAR,
      CONSTRAINT "UQ_generate-entity-id" UNIQUE ("code")
    );
  ```
  Locale codes are necessary to ensure the user selects a locale that exists (no typos, etc.) and to ensure translations relations. These locale codes may be copied from a good list. Strapi has a good one [here](https://github.com/strapi/strapi/blob/main/packages/plugins/i18n/server/constants/iso-locales.json). 
  [Here](https://github.com/vendure-ecommerce/vendure/issues/356) is the discussion the Vendure developers had about locales.
2) Add a default `locale_code` to the store table.
Set to `en` by default but can be changed by an environment variable at setup time, and/or through the admin frontend. This default locale will indicate the locale used for the default fields already available in the current database tables.
3) Add a `store_locales` table, analogous to the `store_currencies` table: this table contains the locales 
available to the store.
  ```sql
  CREATE TABLE "store_locales" (
      "store_id" VARCHAR,
      "locale_code" VARCHAR,
      CONSTRAINT "FK_generate-entity-id" UNIQUE "store_id" REFERENCES "store"("id"),
      CONSTRAINT "FK_generate-entity-id" UNIQUE "locale_code" REFERENCES "locale"("code"),
      PRIMARY KEY ("store_id", "locale_code")
    );
  ```
4) Now create tables that should contain translations. These tables use the parent table id and `locale(id)` as primary key. When a locale code is available in the `store_locales` table, these translations may be available to select. All of these tables should `ON DELETE CASCADE` on the parent table id.
- "product_translations"("product_id","locale_code","title","subtitle","description","handle","material")
- "product_tag_translations"("product_tag_id","locale_code","name")
- "product_type_translations"("product_type_id","locale_code","name")
- "product_variant_translations"("product_variant_id","locale_code","title","material")
- "product_option_value_translations"("product_option_value_id","locale_code","name")
- "product_option_translations"("product_option_id","locale_code","title")
- "product_category_translations"("product_category_id","locale_code","name","handle")
- "product_collection_translations"("product_collection_id""locale_code","title","handle")
- "shipping_option_translations"("shipping_option_id","locale_code","name")
5) Create tables for translated resources. These are, for example, images that may contain text in a different language. All of these tables should `ON DELETE CASCADE` on the parent table id.
- "image_translations"("image_id","locale_code","url")

I think I listed all the fields that contain data shown to customers.

**Reply by dfernandesbsolus** [reply]
@Coachonko we can to do this.

**Reply by NoahPerez** [reply]
@Coachonko Where you able to finish this  push this to production?

@dfernandesbsolus are you doing the backend translation somewhere?

**Reply by einar-hjortdal** [reply]
@NoahPerez I am not knowledgeable about TypeORM therefore I cannot implement these changes in Medusa. I only know raw SQL and vanilla JS. 
I merely proposed changes to the database schema and support them with a rationale, if the community and the developers of Medusa are in agreement, these changes could be introduced by somebody who does know TypeORM.
I will not implement these changes, I will not learn any ORM. 
Medusa seems like a great project, but some things seem to have been implemented terribly, like authn/authz (which fortunately is being overhauled now)

**Reply by samuelstroschein** [reply]
I highly recommend using [BCP-47](https://en.wikipedia.org/wiki/IETF_language_tag) language tags as locales because a) it's the web's current best practice and b) ensures interop.

**Reply by einar-hjortdal** [reply]
> I highly recommend using [BCP-47](https://en.wikipedia.org/wiki/IETF_language_tag) language tags as locales because a) it's the web's current best practice and b) ensures interop.
@samuelstroschein please refer to the strapi and vendure links in my post regarding this.

**Reply by samuelstroschein** [reply]
@Coachonko yeah, they are having a discussing they don't need to have. Use BCP-47 language tags instead of creating your own locale set to follow best practices, the recommendation of W3C, ensure interop between external services and ultimately fulfill user needs. BCP-47 is comprehensive for a reason.

**Reply by einar-hjortdal** [reply]
I believe their concern is valid, perhaps you have not considered how to actually implement your suggestion.
> Adding all variations seems like a bad idea, e.g. according to https://datahub.io/core/language-codes#resource-ietf-language-tags there are over 100 variations of English and over 40 variations of French. In practice, a single "English" version of anything would probably suffice (although I say this as an English speaker from England, so perhaps my perspective is too narrow). In any case, I'm pretty sure that listing 100 variations of English in UI menus is not a desirable solution.

Having all possible BCP-47 language tags in a table would result in a giant table and bad performance, the only other options are:
1. Only support one language and let the users add a new language using their desired tag, without enforcing the subset.
2. Enforce a subset.

Observing how other large and successful projects have solved this issue is the best starting point we have to prevent unnecessary complications

**Reply by samuelstroschein** [reply]
The concern expressed is not about storing language tags but rather a UI question: 

> In any case, I'm pretty sure that listing 100 variations of English in UI menus is not a desirable solution.

Regardless, Medusa can start with support for language codes only (without subtags) but should plan to be compliant with BCP-47 eventually to follow industry best practices (which exist for a reason). The naming of `locales` instead of `languageTags` would already deviate from BCP-47.   

---

> Having all possible BCP-47 language tags in a table would result in a giant table and bad performance [...]

There is likely no need to store all combinations of language tags. Even if, let there be a million rows with 3-4 MB in data, shouldn't be an issue for SQLite.

**Reply by einar-hjortdal** [reply]
> There is likely no need to store all combinations of language tags.

So do you advocate for a subset of tags? I agree a subset is the best option. Can you provide a viable subset? I am not willing to write one, it is extremely tedious. I am unaware of any existing one that could be used.

> Even if, let there be a million rows with 3-4 MB in data, shouldn't be an issue for SQLite.

Please understand that in production Medusa recommends Postgres and not SQLite. Coupled with the slower lookup times, network latency and data transfer will for certain cause bad performance.

**Reply by samuelstroschein** [reply]
> So do you advocate for a subset of tags? 

Yes. Just ensure that the tags are valid BCP-47 language tags. 

I don't have a list. The list from Strapi & Co. is likely OK. I would ensure that the locales contained in those lists are valid tags in the [language tag registry](https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry) and gradually add tags as required.

**Comment by luismelloleite** [comment]
I am willing to translate the project into Portuguese. Let me know if it is ever needed.

**Comment by rico-c** [comment]
Is there an expected time we can use this feature? Need it for sure

**Comment by mj-mehdizadeh** [comment]
I added react.i18next to "admin-ui" and update most of the keywords to support i18n, I created #4190 pull request.

**Reply by dfernandesbsolus** [reply]
Thanks @mj-mehdizadeh maybe we need to split this discussion in 2 topics, one for frontend translations and one for backend (products/attributes).

**Comment by huotarih** [comment]
Interested in this so following.
Would be nice to have a multi-language option for the products and categories and everything basically.

**Comment by boardmain** [comment]
Following this thread ... multi-language is the only thing I'm waiting to start using medusa with my customers....

**Reply by herawais** [reply]
same here, lets collaborate and add this feature.

**Reply by KarthikS373** [reply]
Well I am also interested in collaborating and working on this feature

**Comment by emilselroos** [comment]
I suggest checking out how PrestaShop handles translating in their admin panel UI. In my opinion, they have the best usability for that because they make it very easy to switch between different language versions. Compared to WooCommerce, for example, it's way more simple and saves a lot of time for store managers.

**Comment by gempain** [comment]
Far from perfect but if someone had the patience to extract all strings from the storefront starter to a json file and replace their usages by something like `useTranslations`, we could leverage `next-intl` and product metadata to do the job.

I've written this:

```ts
export function applyMetadataTranslations<
  T extends { metadata?: Record<string, any> | null },
>(item: T, locale: string) {
  const metadata = item.metadata;
  if (!metadata) {
    return;
  }
  console.log(metadata);
  Object.entries(metadata)
    .filter(([key]) => key.endsWith(`_${locale}`))
    .forEach(([key, value]: any) => {
      const prop = key.replace(`_${locale}`, '') as keyof PricedProduct;
      (item as any)[prop] = value;
    });
}
```

which I use this way:

```ts
async function getProducts(
  handle: string,
  locale: string,
): Promise<{ products: PricedProduct[] }> {
  // ... 

  res.body.products.forEach((product: PricedProduct) => {
    applyMetadataTranslations(product, locale);
  });

  return res.body;
}
```

and in my product metadata I have `description_de` with the German translation, `description_fr` and so forth.

It's a bit tedious but hey, it works 😄 I just don't have the time at the moment to extract and translate every string from the starter, and I think the Medusa team has been clear on the subject unfortunately:
- https://github.com/medusajs/nextjs-starter-medusa/issues/109#issuecomment-1470101324
- https://medusajs.com/blog/medusa-contentful-headless/

Next step would be to translate the admin dashboard, and it would be relatively the same task of going through every component and translating it. There seems to have been [attempts](https://github.com/medusajs/medusa/pull/4190) at this already.

**Reply by SGFGOV** [reply]
the only caveat being, next-intl doesn't currently support static paths with SSR.

**Reply by gempain** [reply]
Isn't it in beta ?

**Comment by halfama** [comment]
Looks that this is by far the most requested feature right now. Any chance to see this coming soon?

**Reply by olivermrbl** [reply]
We are currently working on tooling, that will allow you to add translations to any resource in our core. It will not be natively supported, but shipped as an example in a plugin using a couple of new concepts, that we are excited to share soon.

It will be your job to take this example and rework it to fit your use case.

You can expect this to ship later this month or beginning of next.

**Comment by AimanAlmureish** [comment]
If there is a json file we can also create an Arabic one. Now I will be using i18 library to translate it into Arabic

**Comment by olivermrbl** [comment]
Hi everyone. 

Multi-language support for Medusa Admin will be shipping this week in https://github.com/medusajs/medusa/pull/4962. I am aware that this discussion has been split into two - this one being for entity translations - but I wanted to give a quick update for those still tracking admin translations here.

**Comment by yersultanur** [comment]
@olivermrbl Hi, you have done great job. 

I installed fresh medusa backend, but no changes are shown when I try to change the language at the admin panel.

**Comment by yersultanur** [comment]
`
export const supportedLanguages = [
  {
    locale: "en",
    name: "English",
  },
  {
    locale: "fr",
    name: "Français",
  },
  {
    locale: "pt",
    name: "Português (Brasil)",
  },
]

const backendUrl = window.location.origin
const adminPath = process.env.ADMIN_PATH ? `${process.env.ADMIN_PATH}` : "/"

const pathToLoadFrom = `${backendUrl}${adminPath}public/locales/{{lng}}/{{ns}}.json`
`

found the issue with translation you added extra / lower original code, it gives errors in console

`export const supportedLanguages = [
  {
    locale: "en",
    name: "English",
  },
  {
    locale: "fr",
    name: "Français",
  },
  {
    locale: "pt",
    name: "Português (Brasil)",
  },
]

const backendUrl = window.location.origin
const adminPath = process.env.ADMIN_PATH ? `${process.env.ADMIN_PATH}/` : "/"

const pathToLoadFrom = `${backendUrl}${adminPath}public/locales/{{lng}}/{{ns}}.json`
`

**Reply by samuelnunex** [reply]
top

**Comment by RamonEbneter** [comment]
any updates on this? being able to translate products is a must feature and its holding me back from using medusa.
besides that medusa is just amazing! :)

**Reply by RamonEbneter** [reply]
@olivermrbl Any Statement on this one? Will this ever be on the roadmap?

**Reply by olivermrbl** [reply]
We are currently working on tooling, that will allow you to add translations to any resource in our core. It will not be natively supported, but shipped as an example in a plugin using a couple of new concepts, that we are excited to share soon.

It will be your job to take this example and rework it to fit your use case. 

You can expect this to ship later this month or beginning of next.

**Comment by samuelnunex** [comment]
Show

**Comment by luluhoc** [comment]
The localization for categories, products, etc is a must-have..., right now there is almost none of the e-commerce platforms that have localization and multi-currency checkout are available worldwide.

**Reply by olivermrbl** [reply]
We are currently working on tooling, that will allow you to add translations to any resource in our core. It will not be natively supported, but shipped as an example in a plugin using a couple of new concepts, that we are excited to share soon.

It will be your job to take this example and rework it to fit your use case.

You can expect this to ship later this month or beginning of next.

**Reply by luluhoc** [reply]
Great! Thanks for the reply I think I will be using metadata fields for now to translate my store but some of the modules don't support metadata like categories preview feat.

**Reply by pablobarcos** [reply]
@olivermrbl Is there any news about it?

**Comment by yersultanur** [comment]
@olivermrbl Hi, is there any updates on this, waiting for this feature

**Comment by 1lyad** [comment]
@olivermrbl Yo man. Any updates?

**Comment by earmellin** [comment]
Hello, is there any news about this update?

**Reply by SalahAdDin** [reply]
Two years and yet there is not support.

**Reply by TheGP** [reply]
@SaadBazaz 3

**Comment by gtomaka** [comment]
![medusa-multilanguage](https://github.com/medusajs/medusa/assets/34004942/0d6b6b30-2557-49e6-b4ca-33b2716a17d1)
/

Hi, I'm posting a plugin I created. I had a case where the project needed the ability to easily translate product data, but no PIM or CMS was needed. Initially, I thought about writing a native feature for Medusa, but after doing some research, I decided to integrate with Tolgee.

I think it's great software that fits very well into the Medusa stack as it's also open source. Thanks to the integration, I was able to deliver more value much faster with tolgee features such as AI translation assistance or collaborative workflows.

You can read more about this plugin, its features and implementation status in the repo description.

Github repo: https://github.com/rigby-sh/medusa-multilingual-tolgee
NPM: https://www.npmjs.com/package/medusa-multilingual-tolgee
Tolgee: https://tolgee.io

Maybe someone here will also find it useful. 🙌

**Reply by SalahAdDin** [reply]
better if we don't need any extra service to pay.

**Reply by haf** [reply]
<img width="1115" alt="image" src="https://github.com/user-attachments/assets/79329b74-aa8b-478d-b9c5-e068bbf5754c">

**Reply by SalahAdDin** [reply]
> <img alt="image" width="1115" src="https://private-user-images.githubusercontent.com/193115/361148219-79329b74-aa8b-478d-b9c5-e068bbf5754c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ1MzA5NDgsIm5iZiI6MTcyNDUzMDY0OCwicGF0aCI6Ii8xOTMxMTUvMzYxMTQ4MjE5LTc5MzI5Yjc0LWFhOGItNDc4ZC1iOWM1LWUwNjhiYmY1NzU0Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyNFQyMDE3MjhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMDhhOTI1ZDI0YzAzYmY4ZDAwNDdkYWNjM2Y3YjI2YWU3MDEzODU4OWFlOTJiY2M3YzQ0Mzc1Y2YxNmY4ODVkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.u0ZqRfn7ApXuEPeboYw1eYqtovvbM3Rx8-OqBZ7t7Hs">

Yet we have to pay, there is a fee, and a fee system.

**Comment by czw333221** [comment]
> ![medusa-multilanguage](https://github.com/medusajs/medusa/assets/34004942/0d6b6b30-2557-49e6-b4ca-33b2716a17d1)
> /
> 
> Hi, I'm posting a plugin I created. I had a case where the project needed the ability to easily translate product data, but no PIM or CMS was needed. Initially, I thought about writing a native feature for Medusa, but after doing some research, I decided to integrate with Tolgee.
> 
> I think it's great software that fits very well into the Medusa stack as it's also open source. Thanks to the integration, I was able to deliver more value much faster with tolgee features such as AI translation assistance or collaborative workflows.
> 
> You can read more about this plugin, its features and implementation status in the repo description.
> 
> Github repo: https://github.com/rigby-sh/medusa-multilingual-tolgee
> NPM: https://www.npmjs.com/package/medusa-multilingual-tolgee
> Tolgee: https://tolgee.io
> 
> Maybe someone here will also find it useful. 🙌

Hello, is it possible to add the function of translating images, different images for different languages?

**Reply by drloloto** [reply]
Well done!!!

**Comment by matteobaro** [comment]
Any news on multilingual feature??
i think is one of the most important feature !
@MedusaNick

**Reply by TheGP** [reply]
They blocked PR with such functionality below.

**Comment by adriablancafort** [comment]
We would greately appreciate that the Medusa team tackles the i18n feature.
From my point of view, the translatable fields (like product title, description, etc) should be able to be stored in the database for every language the store is translated to.

**Reply by alsherif-khalaf** [reply]
This will make Medusa great

**Reply by SalahAdDin** [reply]
Not in plans yet, right?

**Reply by MrCasquette** [reply]
This would require separating translatable content from content such as images and making translatable content a separate table via a many-to-any relationship. This could make the database more complex. I don't know if it should be done via an external module. For example, a reference table for each element that can be translated via a key. A language table, and finally the translations of each key with each language. In other words, three additional tables that don't complicate the initial database. What do you think?

**Reply by pepijn-vanvlaanderen** [reply]
Creating the database structure(s) and extending API routes in a separate module (we have done this for v1, similar to https://github.com/medusajs/medusa/discussions/2003#discussioncomment-5625663) is fairly straightforward to make in a custom module. However extending the Medusa Admin UI should really be done by the core team to make translating an intuitive part of the UI, instead of adding widgets in places where the original content isn't located.

**Comment by eugenepro2** [comment]
@olivermrbl  do you have this feature in roadmap?

That's really need!)

**Reply by earmellin** [reply]
This is the only feature missing from Medusa that is holding us back from adopting the platform...

**Reply by TheGP** [reply]
Nope and they blocked PR with such functionality implementation too

**Comment by julianlangweb** [comment]
Any Updates on this? 
this seriously is the only reason why i hesitate using medusa

**Reply by earmellin** [reply]
same for us... I don't understand how Medusa can leave behind such an important aspect of the platform

**Reply by MrCasquette** [reply]
If you need a translation you can make a PR. I personnally made the french version. :)

**Reply by julianlangweb** [reply]
hey @triplecasquette its not that i need a different language its that i need multi language support.
like giving a product a french, english, german title, description, etc. and then can show in the frontend the translation of it. 
it's a feature every e-commerce plattform has

**Reply by noureddine-as** [reply]
Same here, I really hope this will come sooner than later, Medusa is JUST AMAZING in every way, except this single aspect.

**Reply by CherryNerd** [reply]
@triplecasquette I don't want my newly made store to have to rely on the Google Translate for example for the other product names and descriptions.
I've seen on Etsy and other sites how disastrous that can be

**Reply by MrCasquette** [reply]
I understand... 🤔 There should be a possibility, but it's clearly Medusa's responsibility to add this major feature. Hope you will all find any solution.

**Comment by e-Naeim** [comment]
I hope this is considered as it's very important and hold us from using the platform :)

**Reply by earmellin** [reply]
same for us...

**Reply by TheGP** [reply]
Nope, they actually rejected the PR below

**Comment by alex-merget** [comment]
any updates here?

**Reply by TheGP** [reply]
They rejected PR with implementation via meta fields below

**Comment by pepijn-vanvlaanderen** [comment]
Currently trying to get this PR https://github.com/medusajs/medusa/pull/13190 through, which could enable some basic translating through metadata fields on Product / Product Variant / Product Collection.

**Reply by TheGP** [reply]
Quick update: they rejected it, nice 💩

**Reply by MedusaNick** [reply]
We are working on a solution to expose a hook for this soon. See comments in the PR.

**Comment by TheGP** [comment]
Hello? It feels like you have agreement with cloud CMS thats why you are blocking such functionality

**Comment by olivermrbl** [comment]
Hi all, a short update on this feature: we have started scoping this week, and expect to have an experimental version ready before the Holidays. Looking forward to sharing this with you all :)

**Reply by TheGP** [reply]
Finally

**Reply by gempain** [reply]
@olivermrbl would this cover admin and storefront ?

**Reply by julianlangweb** [reply]
Amazing, really need this feature

**Reply by olivermrbl** [reply]
@julianlangweb, please see https://github.com/medusajs/medusa/discussions/2003#discussioncomment-15488021.

**Comment by olivermrbl** [comment]
To everyone that have requested localization support in Medusa, I wanted to share that we recently introduced Translations. 

Please refer to our announcement post [here](https://medusajs.com/blog/introducing-experimental-translations/) for more details and use the [setup guide in our documentation](https://docs.medusajs.com/resources/commerce-modules/translation#content) to get started. Bear in mind the feature is still experimental, so the APIs may change. We expect to release the first stable version of Translations later this month.

As always, feedback is welcomed, so please share your thoughts once you've had a chance to try it out.

**Reply by Bernix01** [reply]
Will admin support i18n too?
